test(linter): improve sorting diagnostics

overlookmotel opened the pull request:
Improve (I think) ordering of diagnostic in test snapshots.

1. Sort by filename first, then position. I think this was probably the original intention.
2. Include start column and end line+column in sort, for more deterministic order.
3. Use `sort_by_cached_key`. `Info::new` creates up to 3 `String` allocations, so preferable to cache the result of the closure, to avoid making those allocations repeatedly. `sort_by_key` will likely call the closure many times for each element.

If I've understood right, this reporter is only used in tests, so this PR makes no difference to users.

overlookmotel commented:
* **#12504** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12504?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12504?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12504).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented on apps/oxlint/src/output_formatter/default.rs:
```diff
@@ -132,8 +132,10 @@ mod test_implementation {
 
-            self.diagnostics.sort_by_key(|diagnostic| Info::new(diagnostic).filename);
-            self.diagnostics.sort_by_key(|diagnostic| Info::new(diagnostic).start.line);
+            self.diagnostics.sort_by_cached_key(|diagnostic| {
```
also include the msg inside of the key? just for the case we're using for JS plugins where the loc, file name ect is the same, but the message is diff?

overlookmotel replied:
We were writing at same time. OK, I'll try that and see if it changes the order anywhere else.

overlookmotel replied:
I found I had to include `rule_id` too to get desired order in the JS plugins snapshots.

overlookmotel committed:
test(linter): improve sorting diagnostics (#12504)

Improve (I think) ordering of diagnostic in test snapshots.

1. Sort by filename first, then position. I think this was probably the original intention.
2. Include start column and end line+column in sort, for more deterministic order.
3. Use `sort_by_cached_key`. `Info::new` creates up to 3 `String` allocations, so preferable to cache the result of the closure, to avoid making those allocations repeatedly. `sort_by_key` will likely call the closure many times for each element.

If I've understood right, this reporter is only used in tests, so this PR makes no difference to users.
